pyscf/neo/ase.py

```diff
@@ -86,52 +101,85 @@ def calculate(self, atoms=None, properties=['energy', 'forces', 'dipole'],
+                mf.mf_elec.nlc = 'VV10'
+                mf.mf_elec.grids.prune = None
+                mf.mf_elec.nlcgrids.atom_grid = (50,194)
+                mf.mf_elec.nlcgrids.prune = dft.gen_grid.sg1_prune
+            if isinstance(self.parameters.init_guess, str):
+                if self.parameters.init_guess != 'last_step' and self.parameters.init_guess != 'elec_only':
+                    mf.init_guess = self.parameters.init_guess
+            if self.parameters.conv_tol is not None:
+                mf.conv_tol = self.parameters.conv_tol
+            if self.parameters.conv_tol_grad is not None:
+                mf.conv_tol_grad = self.parameters.conv_tol_grad
+            if self.parameters.add_d3:
+                if not DFTD3_AVAILABLE:
+                    raise RuntimeError('DFTD3 PySCF extension not available')
+                mf.mf_elec = dftd3.dftd3(mf.mf_elec)
+            if self.parameters.add_solvent:
+                mf.scf(cycle=0) # TODO: remove this
+                mf = mf.ddCOSMO()
+
+            # always use dict as the initial guess
+            if isinstance(self.parameters.init_guess, dict):
+                dm0 = self.parameters.init_guess
+                mf.scf(dm0=[dm0['e']]+dm0['n'])
+            else:
+                mf.scf()
+            self.mf = mf
         else:
-            mf = neo.CDFT(mol, unrestricted=True, epc=self.parameters.epc)
-        mf.mf_elec.xc = self.parameters.xc
-        if self.parameters.atom_grid is not None:
-            mf.mf_elec.grids.atom_grid = self.parameters.atom_grid
-        if self.parameters.add_vv10:
-            mf.mf_elec.nlc = 'VV10'
-            mf.mf_elec.grids.prune = None
-            mf.mf_elec.nlcgrids.atom_grid = (50,194)
-            mf.mf_elec.nlcgrids.prune = dft.gen_grid.sg1_prune
-        if self.parameters.init_guess is not None:
-            mf.init_guess = self.parameters.init_guess
-        if self.parameters.conv_tol is not None:
-            mf.conv_tol = self.parameters.conv_tol
-        if self.parameters.conv_tol_grad is not None:
-            mf.conv_tol_grad = self.parameters.conv_tol_grad
-        if self.parameters.add_d3:
-            if not DFTD3_AVAILABLE:
-                raise RuntimeError('DFTD3 PySCF extension not available')
-            mf.mf_elec = dftd3.dftd3(mf.mf_elec)
-        if self.parameters.add_solvent:
-            mf.scf(cycle=0) # TODO: remove this
-            mf = mf.ddCOSMO()
+            # if mf is there, reset it with new mol
+            self.mf.reset(mol)
+            if self.parameters.init_guess == 'last_step':
+                self.mf.scf(dm0=[self.rdm1] + self.dm_nuc)
+            elif self.parameters.init_guess == 'elec_only':
+                self.mf.init_guess = self.rdm1
+                self.mf.scf()
+            elif isinstance(self.parameters.init_guess, dict):
+                dm0 = self.parameters.init_guess
+                self.mf.scf(dm0=[dm0['e']]+dm0['n'])
+            else:
```

Comment:
What about other initial guess conditions? `mf.init_guess = self.parameters.init_guess` choice is missing here.

Reply:
The rest two initial guess options are "last step".  This piece of code, is the first time the calculator is called. There are no previous SCF results for them to be the initial guess.

Reply:
https://github.com/theorychemyang/pyscf/blob/28e8f5cb910f31444b9018e30131c7e0f0113d3a/pyscf/neo/ase.py#L54C22-L54C22
I added `init_guess` to take string inputs to directly pass to pyscf, you don't need to implement it, I already did it. What I am saying is that your code here broke the functionality, for example, if I want the `huckel` initial guess. Your if else did not check the str case in this else block, though you do have the proper check in the previous if block.

Reply:
> ~The rest two initial guess options are "last step". This piece of code, is the first time the calculator is called. There are no previous SCF results for them to be the initial guess.~
> 
> For the other initial guess,` mf.init_guess` is initialized the first time the calculator is called, and will not be modified in the future step. There is no need to assign its value every time.

Better not assuming that.

Reply:
Or you can assume that, `mf.init_guess` is indeed not changed, but change the `parameters.init_guess` parsing code here such that it is not parsed again (from the initialization you already knew what the initial guess is, right?)

Reply:
Please read updated comment and my latest comment.

Reply:
Yes I know.
>There is no need to parse parameters.init_guess in this else branch.

but your code here in the `else` branch DID parse it again. I know it is 2nd and later steps, but the point is, ideally it should be parsed for only one time.

Or put it in that way, the code does not need the big if else at all.

TBH it is the problem of the calculator here that the initialization did not occur in the initializer but incorrectly in the `calculate` function . I need to refactor it first, then you can change according to the updated `ase.py` and file a new PR. Closing for now.